skel.1
.\"
.\" SKEL.1 --Manual page for "skel", a skeleton file expander.
.\"
.TH SKEL 1 "DATE" "VERSION" "The Other Manual"
.SH NAME
skel \- create some folders and files based on a skeleton template
.SH SYNOPSIS
.SY skel
.OP --list
.YS
.SY skel
.OP --name myname
.OP --script filter
.OP --include path:path...
.OP --force
.OP --camel
.OP --windows
.I skeleton \&.\|.\|.\&
.YS
.SH DESCRIPTION
The
.B skel
command creates one or more files from a skeleton template file.
The skeleton is a self-extracting shell archive, which is transformed
as it is expanded to replace the word "skeleton" with the specified
.IR name .
The replacement is case sensitive in the following way; if the
replacement name is (e.g.) "foo_bar":
.PP
.RS 2
.PD 0
.TP 14
.B skeleton
is replaced with foo_bar (or fooBar)
.TP
.B Skeleton
is replaced with FooBar
.TP
.B SKELETON
is replaced with FOO_BAR.
.RE
.PD
.PP
Note that this transformation is applied to the shell archive as a
whole, so it applies equally to the files' names and to their
contents.
.PP
The
.I --camel
option controls whether
.I name
is translated as
.I foo_bar
or
.I fooBar.
.PP
Binary files unpacked by
.B skel
are assumed to be
.BR zip (1)
archives: it applies
.B unzip
to the file, and if that succeeds, it applies the replacement
transformation to the expanded files and then re-zips the files back
into the archive.
This strategy works well for Microsoft Office
files, which are
.B zip
archives of XML documents.
.PP
By default,
.B skel
searches for skeletons in the directory
.IR /usr/local/share/skel.
although this may be overridden by the environment variable
.B SKELPATH
or the
.B --include
command line option, as described below.
.SH OPTIONS
.TP
.BR \-l ,\  \-\-list
List the available skeleton templates.
.TP
.BI \-n\  name
.TQ
.BI \-\-name\fR= name
Modify the generated file names and contents based on
.IR name .
.TP
.BI \-s\  script
.TQ
.BI \-\-script\fR= filter-script
Further modify the generated content by a
script, a simple filter that reads text from
.I stdin
and prints the filtered/modified output to
.IR stdout .
If the environment variable SKEL_FILTER is defined, its value will be used to
name the script.  The script must be named as either an absolute path name, or
name an executable command in PATH.

For example, the script can be used to customise generic copyright
information, using sed:

.EX
#!/usr/bin/env -S sed -f
s/\\<COPYRIGHT\\>/Copyright (c) YEAR AUTHOR/g
s/\\<YEAR\\>/2038/g
s/\\<AUTHOR\\>/Bozo the Clown/g
.EE
.TP
.BI \-I\  path-spec
.TQ
.BI \-\-include\fR= path-spec
Specify the directories to search for skeleton templates, as a colon-separated
list.
.TP
.BR \-p ,\  \-\-pascal
Control expansion of
.I name
to
.IR pascalCase .
.TP
.BR \-f ,\  \-\-force
Force overwriting of existing files when expanding the skeleton template.
.TP
.BR \-w ,\  \-\-windows
Text files are output with Windows format (i.e. <cr><lf>) line endings.
.PD
.SS "Logging Options"
The
.B skel
command specifies its internal logging level in a
.BR syslog (3)
compatible way.  The default logging level is "notice".
The following options set the logging to a specific level.
.PP
.PD 0
.TP 18
.BR \-q ,\  \-\-quiet
Set the logging level to "warning".
.TP
.BR \-v ,\  \-\-verbose
Set the logging level to "info".
.TP
.BR \-_ ,\  \-\-debug
Set the logging level to "debug".
.TP
.BI \-\-log-level\fR= level
Set the logging level to the specified syslog level.
.PD
.SH ENVIRONMENT
.B skel
is sensitive to the following environment variables:
.PD 0
.TP 14
SKELPATH
The default search path for
.B skel
skeleton templates, as a colon-separated list of directories.
.TP
LOG_LEVEL
Specifies the default logging level.
If not defined, the logging level "notice"
is used.
.TP
MIDDENPATH
The search path for additional shell library functions.
.TP
SKEL_FILTER
The name of a filter script to make further changes to the skeleton.
.PD
.SH "EXIT STATUS"
.B Skel
exits with the following status:
.PD 0
.TP 4
0
Successfully processed all skeleton templates.
.TP
1
Unable to process all skeleton templates (e.g. cannot open skeleton file).
.TP
2
There was an error in the command line arguments (e.g. unrecognised option).
.SH EXAMPLES
.EX
.B $ skel -l
some output...
.EE
.SH BUGS
The template name transformation is done via a blind
.BR sed ;
there is a possiblility that binary files will be corrupted.
.SH SEE ALSO
.BR skel-shar (1),
.BR shar (1),
.BR sed (1),
.BR zip (1),
.BR logger (1).


skel.sh
#!/bin/sh
#
# SKEL --Create files from a skeleton.
#
# Contents:
# main()           --Process cmd-line options etc.
# usage()          --echo this script's help message.
# list_skeletons() --List the skeletons in skel's include path.
# fill_skeleton()  --Expand a skeleton file, replacing template variables etc.
# skel_binary()    --Expand a binary skeleton file, assumed to be a zip.
#
# Remarks:
# This utility creates file(s) from supplied templates.  The templates
# are ".sha" shell archives; the template file undergoes keyword
# expansion before the archive is unpacked.  This allows both the name
# and the content to manipulated in a uniform way.
#
. "midden"
require "log"
require "getopt"
require "wordy"

version=
include=${SKELPATH:-/usr/local/share/skel}
script=${SKEL_FILTER:-cat}
name=
opts="f.force;I.include=$include;l.list;n.name=$name;c.camel;s.script=$script;V.version;w.windows;?.help"
opts="$opts;$LOG_GETOPTS"
sh_opts=

#
# main() --Process cmd-line options etc.
#
main()
{
    if ! eval $(getopt_long_args -d "$opts" "$@"); then
	usage "$opts" >&2
	exit 2
    fi
    if [ "$help" ]; then
	usage "$opts"
	exit 0
    fi
    if [ "$version" ]; then
        printf 'skel version %s\n' "VERSION"
        exit 0
    fi

    log_getopts
    if [ "$list" ]; then
	list_skeletons
	exit 0
    fi

    if [ "$#" -eq 0 ]; then
	log_quit "no template specified"
    fi
    if [ "$name" ]; then
	name=$(snake_case "$name")
    else
	log_quit "no name specified"
    fi
    if ! filter=$(which "$script"); then
	log_quit 'cannot execute file "%s"' "$script"
    fi

    # REVISIT: read templates from stdin if no-args?
    for file; do
	local skel_path=$(resolve_path "$include" "$file.sha")

	if [ -f "$file" ]; then
	    skel_path=$file
	fi

	if [ ! "$skel_path" ]; then
	    err 'no such skeleton: "%s"' "$file"
	    continue
	fi
	fill_skeleton "$skel_path"
	# TODO: apply and remove post-processing script
    done
}

#
# usage() --Echo this program's help message.
#
usage()
{
    printf 'skel version %s\n' "$version"
    getopt_usage "skel [options] -n name skeleton..." "$1"
    cat <<-EOF
	e.g.
	skel -n my_new_project c-project"
	EOF
}

#
# list_skeletons() --List the skeletons in skel's include path.
#
list_skeletons()
{
    local IFS=:

    for dir in $include; do
        debug 'list %s' "$dir"
        if [ -d "$dir" ]; then
	    find "$dir" -name "*.sha" | sed -e "s|$dir/||;s/[.]sha//g"
	fi
    done | sort -u
}

#
# fill_skeleton() --Expand a skeleton file, replacing template variables etc.
#
# Remarks:
# The skeleton files are shell archives.
#
fill_skeleton()
{
    local status=
    local skel_file=$1 tmp_file=skel-$$.sh idx_file="skel-idx-$$.txt"
    local pascal_name="$(pascal_case $name)"
    local upper_name="$(echo $name | tr a-z A-Z)"
    local plural_name=$(plural $name)
    local plural_pascal_name="$(pascal_case $plural_name)"
    local plural_upper_name="$(echo $plural_name | tr a-z A-Z)"

    if [ "$camel" ]; then
	name="$(camel_case $name)"
    fi
    local transform="s/skeletons/$plural_name/g;s/Skeletons/$plural_pascal_name/g;s/SKELETONS/$plural_upper_name/g"
    local transform="$transform;s/skeleton/$name/g;s/Skeleton/$pascal_name/g;s/SKELETON/$upper_name/g"

    info 'loading skeleton "%s"' "$skel_file"
    debug 'name: "%s", "%s", "%s"' "$name" "$pascal_name" "$upper_name"
    debug 'plural.name: "%s", "%s", "%s"' "$plural_name" "$plural_pascal_name" "$plural_upper_name"

    sed -e "$transform" <"$skel_file" | "$filter" >"$tmp_file"

    sh_opts="${force:+-f} ${windows:+-w}"
    sh "$tmp_file" $sh_opts > "$idx_file"
    status="$?"

    if [ "$status" -eq 0 ]; then
	while read action file type; do
	    info "%s %s" "$file" "$type"
	    if [ "$type" = "binary" ]; then
		skel_binary "$file"
	    fi
	done <"$idx_file"
    fi

    rm "$tmp_file" "$idx_file"
    return "$status"
}

#
# skel_binary() --Expand a binary skeleton file, assumed to be a zip.
#
# Remarks:
# Several applications use a zip-compressed collection of XML files
# to store content.
#
# See Also:
# midden/moxie.shl
#
skel_binary()
{				# TODO: set transform, dammit.
    local file= zip_dir="zip-$$.d"
    for file; do
	mkdir "$zip_dir"
	(
	    cd "$zip_dir"
	    if quietly unzip -q "../$file"; then
		info 'processing zip file %s' "$file"
		log_cmd sed -i -e "$transform" ${script:+-f $script} $(find * -type f)
		zip -q "../$file" $(find * -type f)
	    fi
	)
	rm -rf "$zip_dir"
    done
}

main "$@"
